contracts/Controller.sol

```diff
@@ -122,57 +130,155 @@ contract Controller is Ownable {
-     * @param _operator account operator
-     * @return true if it is an account operator, otherwise false
+     * @notice get number of vaults in a specific account
+     * @param _accountOwner account owner address
+     * @return number of vaults
      */
-    // function isOperator(address _owner, address _operator) public view returns (bool) {
-    // }
+    function getAccountVaultCounter(address _accountOwner) external view returns (uint256) {
+        return accountVaultCounter[_accountOwner];
+    }
+
+    /**
+     * @notice function to check if otoken is expired
+     * @param _otoken otoken address
+     * @return true if otoken is expired, else return false
+     */
+    function isExpired(address _otoken) public view returns (bool) {
+        uint256 otokenExpiryTimestamp = OtokenInterface(_otoken).expiryTimestamp();
+
+        return now > otokenExpiryTimestamp;
+    }
+
+    /**
+     * @notice Return a specific vault.
+     * @param _owner account owner.
+     * @param _vaultId vault.
+     * @return Vault struct
+     */
+    function getVault(address _owner, uint256 _vaultId) public view returns (MarginAccount.Vault memory) {
+        return vaults[_owner][_vaultId];
+    }
 
     /**
      * @notice Execute actions on a certain vault
      * @dev For each action in the action Array, run the corresponding action
      * @param _actions An array of type Actions.ActionArgs[] which expresses which actions the user want to execute.
-     * @return Vault strcut. The new vault data that has been modified (or null vault if no action affected any vault)
      */
-    // function _runActions(Actions.ActionArgs[] memory _actions) internal returns (MarginAccount.Vault memory) {
-    // }
+    function _runActions(Actions.ActionArgs[] memory _actions) internal {
+        for (uint256 i = 0; i < _actions.length; i++) {
+            Actions.ActionArgs memory action = _actions[i];
+            Actions.ActionType actionType = action.actionType;
+
+            uint256 prevActionVaultId;
+
+            if (actionType == Actions.ActionType.OpenVault) {
+                // check if this action is manipulating the same vault as all other actions, other than SettleVault
+                require(
+                    prevActionVaultId == 0 || action.vaultId == prevActionVaultId,
+                    "Controller: can not run actions on different vaults"
+                );
+                prevActionVaultId = action.vaultId;
+
+                _openVault(Actions._parseOpenVaultArgs(action));
+            }
+        }
+    }
+
+    /**
+     * @notice verify vault final state after executing all actions
+     * @param _accountOwner account owner address
+     * @param _vaultId vault id related to the vault that will be checked
+     */
+    function _verifyFinalState(address _accountOwner, uint256 _vaultId) internal view {
+        MarginAccount.Vault memory vault = vaults[_accountOwner][_vaultId];
+
+        if (vault.shortOtokens.length > 0) {
```

Comment:
shouldn't this check be in isValidState function? 

Also, why is this check there? Even if short.length is 0 we should check this right?

Reply:
if short.length is 0 then `vault.shortOtokens[0]` will revert with invalid opcode as we are accessing an invalid index